File: rust/sdk/src/types/transfer.rs
use crate::account::AccountId;
use crate::error::{M10Error, M10Result};
use crate::types::TxId;
use crate::{EnhancedTransfer, EnhancedTransferStep, TransactionExt};
use core::convert::{From, TryFrom};
use core::result::Result;
use core::time::Duration;
use m10_protos::prost::Any;
use m10_protos::sdk::transaction_data::Data;
use m10_protos::{sdk, MetadataExt, MetadataType};
use serde::{Serialize, Serializer};
use serde_with::{serde_as, SerializeAs};
use std::time::{SystemTime, UNIX_EPOCH};

#[serde_as]
#[derive(Clone, Debug, serde::Serialize)]
pub struct TransferStep {
    pub from: AccountId,
    pub to: AccountId,
    pub amount: u64,
    #[serde_as(as = "Vec<AsMetadata>")]
    pub metadata: Vec<Any>,
}

struct AsMetadata;

impl SerializeAs<Any> for AsMetadata {
    fn serialize_as<S: Serializer>(data: &Any, serializer: S) -> Result<S::Ok, S::Error> {
        let metatdata = if let Some(sdk::metadata::Memo { plaintext }) = data.protobuf() {
            format!("Memo: {plaintext}")
        } else if let Some(sdk::metadata::Fee {}) = data.protobuf() {
            "Fee".to_string()
        } else if let Some(sdk::metadata::Withdraw { bank_account_id }) = data.protobuf() {
            format!("Withdraw {bank_account_id}")
        } else if let Some(sdk::metadata::Deposit { bank_account_id }) = data.protobuf() {
            format!("Deposit {bank_account_id}")
        } else if let Some(sdk::metadata::Attachment { object_id, .. }) = data.protobuf() {
            format!("Attachment: {object_id}")
        } else if let Some(sdk::metadata::Contract { endorsements, .. }) = data.protobuf() {
            format!("Contract with {} endorsments", endorsements.len())
        } else if let Some(sdk::metadata::SelfTransfer {
            from_account_name,
            to_account_name,
        }) = data.protobuf()
        {
            format!("Self transfer from {from_account_name} to {to_account_name}")
        } else if let Some(sdk::metadata::RebalanceTransfer {}) = data.protobuf() {
            "Rebalance transfer".to_string()
        } else if let Some(sdk::metadata::TokenWithdraw {}) = data.protobuf() {
            "Tokenization".to_string()
        } else if let Some(sdk::metadata::OfflineTransfer { input }) = data.protobuf() {
            format!("Offline payment from {input}")
        } else {
            "Unknown".to_string()
        };

        serializer.serialize_str(&metatdata)
    }
}

#[cfg(feature = "format")]
impl std::fmt::Display for TransferStep {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            from,
            to,
            amount,
            metadata,
        } = self;
        write!(f, "Step{{ from={from} to={to} amount={amount} metadata=[",)?;
        for m in metadata {
            write!(f, "{},", m.type_url)?;
        }
        write!(f, "] }}")
    }
}

impl MetadataExt for TransferStep {
    fn with_type<M: MetadataType>(&self) -> Option<&[u8]> {
        self.metadata
            .iter()
            .find(|a| a.type_url == M::TYPE_URL)
            .map(|a| a.value.as_slice())
    }
}

impl TryFrom<sdk::TransferStep> for TransferStep {
    type Error = M10Error;

    fn try_from(step: sdk::TransferStep) -> Result<Self, Self::Error> {
        let sdk::TransferStep {
            from_account_id,
            to_account_id,
            amount,
            metadata,
        } = step;
        Ok(TransferStep {
            from: AccountId::try_from_be_slice(&from_account_id)?,
            to: AccountId::try_from_be_slice(&to_account_id)?,
            amount,
            metadata,
        })
    }
}

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[derive(Clone, Debug, Copy, Serialize, PartialEq, Eq)]
pub enum TransferStatus {
    Pending,
    Accepted,
    Rejected,
    Expired,
}

impl From<sdk::finalized_transfer::TransferState> for TransferStatus {
    fn from(status: sdk::finalized_transfer::TransferState) -> Self {
        match status {
            sdk::finalized_transfer::TransferState::Pending => TransferStatus::Pending,
            sdk::finalized_transfer::TransferState::Accepted => TransferStatus::Accepted,
            sdk::finalized_transfer::TransferState::Rejected => TransferStatus::Rejected,
            sdk::finalized_transfer::TransferState::Expired => TransferStatus::Expired,
        }
    }
}

#[serde_as]
#[derive(Clone, Debug, Serialize)]
pub struct Transfer {
    pub tx_id: TxId,
    #[serde_as(as = "serde_with::hex::Hex")]
    pub context_id: Vec<u8>,
    pub timestamp: SystemTime,
    pub steps: Vec<TransferStep>,
    pub success: bool,
    pub status: TransferStatus,
}

#[cfg(feature = "format")]
impl std::fmt::Display for Transfer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            tx_id,
            context_id,
            timestamp,
            steps,
            success,
            status,
        } = self;
        write!(f, "Transfer{{ tx_id={tx_id} context_id={} timestamp={timestamp:?} success={success} status={status} steps=[", hex::encode(context_id))?;
        for step in steps {
            write!(f, "{step},")?;
        }
        write!(f, "] }}")
    }
}

impl TryFrom<sdk::FinalizedTransfer> for Transfer {
    type Error = M10Error;

    fn try_from(transfer: sdk::FinalizedTransfer) -> Result<Self, Self::Error> {
        Ok(Transfer {
            tx_id: transfer.tx_id,
            context_id: transfer.context_id,
            success: transfer.error.is_none(),
            timestamp: UNIX_EPOCH + Duration::from_micros(transfer.timestamp),
            steps: transfer
                .transfer_steps
                .into_iter()
                .map(TransferStep::try_from)
                .collect::<M10Result<_>>()?,
            status: TransferStatus::from(
                sdk::finalized_transfer::TransferState::try_from(transfer.state)
                    .map_err(|_| M10Error::InvalidTransaction)?,
            ),
        })
    }
}

impl TryFrom<sdk::FinalizedTransaction> for Transfer {
    type Error = M10Error;

    fn try_from(txn: sdk::FinalizedTransaction) -> Result<Self, Self::Error> {
        let response = txn.response.as_ref().ok_or(M10Error::InvalidTransaction)?;
        let context_id = txn
            .request
            .as_ref()
            .ok_or(M10Error::InvalidTransaction)?
            .context_id
            .clone();
        let success = response.error.is_none();
        let tx_id = response.tx_id;
        let timestamp = UNIX_EPOCH + Duration::from_micros(response.timestamp);
        match txn.data().ok_or(M10Error::InvalidTransaction)? {
            Data::Transfer(transfer) => Ok(Self {
                tx_id,
                context_id,
                timestamp,
                steps: transfer
                    .transfer_steps
                    .iter()
                    .cloned()
                    .map(TransferStep::try_from)
                    .collect::<M10Result<_>>()?,
                success,
                status: TransferStatus::Accepted,
            }),
            Data::InitiateTransfer(transfer) => Ok(Self {
                tx_id,
                context_id,
                timestamp,
                steps: transfer
                    .transfer_steps
                    .iter()
                    .cloned()
                    .map(TransferStep::try_from)
                    .collect::<M10Result<_>>()?,
                success,
                status: TransferStatus::Pending,
            }),
            Data::CommitTransfer(commit) => {
                let status = if success {
                    TransferStatus::from(
                        sdk::finalized_transfer::TransferState::try_from(commit.new_state)
                            .map_err(|_| M10Error::InvalidTransaction)?,
                    )
                } else {
                    TransferStatus::Pending
                };
                let transfer = response
                    .transfer_committed
                    .as_ref()
                    .ok_or(M10Error::InvalidTransaction)?;
                Ok(Self {
                    tx_id,
                    context_id,
                    timestamp,
                    steps: transfer
                        .transfer_steps
                        .iter()
                        .cloned()
                        .map(TransferStep::try_from)
                        .collect::<M10Result<_>>()?,
                    success,
                    status,
                })
            }
            _ => Err(M10Error::InvalidTransaction),
        }
    }
}

impl MetadataExt for Transfer {
    fn with_type<M: MetadataType>(&self) -> Option<&[u8]> {
        self.steps.iter().find_map(TransferStep::with_type::<M>)
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct ExpandedTransfer {
    pub tx_id: TxId,
    pub context_id: Vec<u8>,
    pub timestamp: SystemTime,
    pub steps: Vec<ExpandedTransferStep>,
    pub success: bool,
    pub status: TransferStatus,
}

#[cfg(feature = "format")]
impl std::fmt::Display for ExpandedTransfer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            tx_id,
            context_id,
            timestamp,
            steps,
            success,
            status,
        } = self;
        write!(f, "Transfer{{ tx_id={tx_id} context_id={} timestamp={timestamp:?} success={success} status={status} steps=[", hex::encode(context_id))?;
        for step in steps {
            write!(f, "{step},")?;
        }
        write!(f, "] }}")
    }
}

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[cfg_attr(
    feature = "format",
    display("Account{{ id={id} name={public_name} image={profile_image_url} currency={code}({decimals}) }}")
)]
#[derive(Clone, Debug, Serialize)]
pub struct ExpandedAccount {
    pub id: AccountId,
    pub public_name: String,
    pub profile_image_url: String,
    pub code: String,
    pub decimals: u32,
}

#[derive(Clone, Debug, Serialize)]
pub struct ExpandedTransferStep {
    pub from: ExpandedAccount,
    pub to: ExpandedAccount,
    pub amount: u64,
    // TODO @sadroeck - fixme
    #[serde(skip)]
    pub metadata: Vec<Any>,
}

#[cfg(feature = "format")]
impl std::fmt::Display for ExpandedTransferStep {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            from,
            to,
            amount,
            metadata,
        } = self;
        write!(f, "Step{{ from={from} to={to} amount={amount} metadata=[",)?;
        for m in metadata {
            write!(f, "{},", m.type_url)?;
        }
        write!(f, "] }}")
    }
}

impl TryFrom<(EnhancedTransferStep, sdk::TransferStep)> for ExpandedTransferStep {
    type Error = M10Error;

    fn try_from(
        (step, raw): (EnhancedTransferStep, sdk::TransferStep),
    ) -> Result<Self, Self::Error> {
        let from = step.from.ok_or(M10Error::InvalidTransaction)?;
        let to = step.to.ok_or(M10Error::InvalidTransaction)?;
        Ok(Self {
            from: ExpandedAccount {
                id: AccountId::try_from_be_slice(&from.account_id)?,
                public_name: from.public_name,
                profile_image_url: from.profile_image_url,
                code: from.code,
                decimals: from.decimal_places,
            },
            to: ExpandedAccount {
                id: AccountId::try_from_be_slice(&to.account_id)?,
                public_name: to.public_name,
                profile_image_url: to.profile_image_url,
                code: to.code,
                decimals: to.decimal_places,
            },
            amount: raw.amount,
            metadata: raw.metadata,
        })
    }
}

impl TryFrom<EnhancedTransfer> for ExpandedTransfer {
    type Error = M10Error;

    fn try_from(transfer: EnhancedTransfer) -> Result<Self, Self::Error> {
        Ok(Self {
            tx_id: transfer.transfer.tx_id,
            context_id: transfer.transfer.context_id,
            timestamp: UNIX_EPOCH + Duration::from_micros(transfer.transfer.timestamp),
            steps: transfer
                .enhanced_steps
                .into_iter()
                .zip(transfer.transfer.transfer_steps.into_iter())
                .map(ExpandedTransferStep::try_from)
                .collect::<M10Result<_>>()?,
            success: transfer.transfer.error.is_none(),
            status: TransferStatus::from(
                sdk::finalized_transfer::TransferState::try_from(transfer.transfer.state)
                    .map_err(|_| M10Error::InvalidTransaction)?,
            ),
        })
    }
}

#[derive(serde::Serialize)]
pub struct TransferView {
    #[serde(flatten)]
    pub transfer: TransferData,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx_sender: Option<String>,
}

#[derive(serde::Serialize)]
#[serde(untagged)]
pub enum TransferData {
    Basic(Transfer),
    Expanded(ExpandedTransfer),
}

File: node/m10-sdk/src/types/transfer.ts
import type { google } from "../../protobufs";
import { m10 } from "../../protobufs";
import { M10Error } from "../error";
import type { TxId } from "../ids";
import { AccountId } from "../ids";
import type { EnhancedTransfer, EnhancedTransferStep } from "../transfer_ext";
import { isNone, unwrap } from "../utils";


export interface ITransferStep {
    from: AccountId;
    to: AccountId;
    amount: LongNumber;
    metadata: google.protobuf.IAny[];
}

export class TransferStep implements ITransferStep{
    public from: AccountId;
    public to: AccountId;
    public amount: LongNumber;
    public metadata: google.protobuf.IAny[];

    public constructor(properties: ITransferStep) {
        this.from = properties.from;
        this.to = properties.to;
        this.amount = properties.amount;
        this.metadata = properties.metadata;
    }

    // TODO: Implement `with_type` function when metadate is a bit more defined
    // impl MetadataExt for TransferStep {
    //     fn with_type<M: MetadataType>(&self) -> Option<&[u8]> {
    //         self.metadata
    //             .iter()
    //             .find(|a| a.type_url == M::TYPE_URL)
    //             .map(|a| a.value.as_slice())
    //     }
    // }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     * @throws  {M10Error.InvalidAccountId}     if invalid account id was used
     */
    public static tryFrom(transferStep: m10.sdk.transaction.ITransferStep): TransferStep {
        return new TransferStep({
            from: AccountId.fromUint8Array(unwrap(transferStep.fromAccountId, M10Error.InvalidTransaction())),
            to: AccountId.fromUint8Array(unwrap(transferStep.toAccountId, M10Error.InvalidTransaction())),
            amount: unwrap(transferStep.amount, M10Error.InvalidTransaction()),
            metadata: unwrap(transferStep.metadata, M10Error.InvalidTransaction()),
        });
    }
}

export enum TransferStatus {
    Pending,
    Accepted,
    Rejected,
    Expired,
}

export interface ITransfer {
    txId: TxId;
    contextId: Uint8Array;
    timestamp: LongNumber;
    steps: TransferStep[];
    success: boolean;
    status: TransferStatus;
}

export class Transfer implements ITransfer {
    public txId: TxId;
    public contextId: Uint8Array;
    public timestamp: LongNumber;
    public steps: TransferStep[];
    public success: boolean;
    public status: TransferStatus;

    public constructor(properties: ITransfer) {
        this.txId = properties.txId;
        this.contextId = properties.contextId;
        this.timestamp = properties.timestamp;
        this.steps = properties.steps;
        this.success = properties.success;
        this.status = properties.status;
    }

    // TODO: Implement `with_type` function when metadate is a bit more defined
    // impl MetadataExt for Transfer {
    //     fn with_type<M: MetadataType>(&self) -> Option<&[u8]> {
    //         self.steps.iter().find_map(TransferStep::with_type::<M>)
    //     }
    // }

    public static getTransferStatus(state: m10.sdk.transaction.FinalizedTransfer.TransferState): TransferStatus {
        switch (state) {
            case m10.sdk.transaction.FinalizedTransfer.TransferState.PENDING:
                return TransferStatus.Pending;
            case m10.sdk.transaction.FinalizedTransfer.TransferState.ACCEPTED:
                return TransferStatus.Accepted;
            case m10.sdk.transaction.FinalizedTransfer.TransferState.REJECTED:
                return TransferStatus.Rejected;
            case m10.sdk.transaction.FinalizedTransfer.TransferState.EXPIRED:
                return TransferStatus.Expired;
        }
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(transfer: m10.sdk.transaction.IFinalizedTransfer): Transfer {
        return new Transfer({
            txId: unwrap(transfer.txId, M10Error.InvalidTransaction()),
            contextId: unwrap(transfer.contextId, M10Error.InvalidTransaction()),
            timestamp: unwrap(transfer.timestamp, M10Error.InvalidTransaction()),
            steps: unwrap(transfer.transferSteps, M10Error.InvalidTransaction()).map(TransferStep.tryFrom),
            success: isNone(transfer.error),
            status: Transfer.getTransferStatus(unwrap(transfer.state, M10Error.InvalidTransaction())),
        });
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFromTxn(txn: m10.sdk.IFinalizedTransaction): Transfer {
        const txnData = new m10.sdk.transaction.TransactionData(unwrap(txn.request?.data, M10Error.InvalidTransaction()));

        const contextId = unwrap(txn.request?.contextId, M10Error.InvalidTransaction());
        const success = isNone(txn.response?.error);
        const txId = unwrap(txn.response?.txId, M10Error.InvalidTransaction());
        const timestamp = unwrap(txn.response?.timestamp, M10Error.InvalidTransaction());

        switch (unwrap(txnData.data, M10Error.InvalidTransaction())) {
            case "transfer":
                return new Transfer({
                    txId: txId,
                    contextId: contextId,
                    timestamp: timestamp,
                    steps: unwrap(txnData.transfer?.transferSteps, M10Error.InvalidTransaction())
                        .map(TransferStep.tryFrom),
                    success: success,
                    status: TransferStatus.Accepted,
                });
            case "initiateTransfer":
                return new Transfer({
                    txId: txId,
                    contextId: contextId,
                    timestamp: timestamp,
                    steps: unwrap(txnData.initiateTransfer?.transferSteps, M10Error.InvalidTransaction())
                        .map(TransferStep.tryFrom),
                    success: success,
                    status: TransferStatus.Pending,
                });
            case "commitTransfer":
                return new Transfer({
                    txId: txId,
                    contextId: contextId,
                    timestamp: timestamp,
                    steps: unwrap(txn.response?.transferCommitted?.transferSteps, M10Error.InvalidTransaction())
                        .map(TransferStep.tryFrom),
                    success: success,
                    status: (
                        success
                            ? Transfer.getTransferStatus(
                                unwrap(txn.request?.data?.commitTransfer?.newState, M10Error.InvalidTransaction()),
                            )
                            : TransferStatus.Pending
                    ),
                });
            default:
                throw M10Error.InvalidTransaction();
        }
    }
}


export interface ExpandedAccount {
    id: AccountId;
    publicName: string;
    profileImageUrl: string;
    code: string;
    decimals: number;
}


export interface IExpandedTransferStep {
    from: ExpandedAccount;
    to: ExpandedAccount;
    amount: LongNumber;
    metadata: google.protobuf.IAny[];
}

export class ExpandedTransferStep implements IExpandedTransferStep {
    public from: ExpandedAccount;
    public to: ExpandedAccount;
    public amount: LongNumber;
    public metadata: google.protobuf.IAny[];


    public constructor(properties: IExpandedTransferStep) {
        this.from = properties.from;
        this.to = properties.to;
        this.amount = properties.amount;
        this.metadata = properties.metadata;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     * @throws  {M10Error.InvalidAccountId}     if invalid account id was used
     */
    public static tryFrom(step: EnhancedTransferStep, raw: m10.sdk.transaction.ITransferStep): ExpandedTransferStep {
        const from: ExpandedAccount = {
            id: AccountId.fromUint8Array(unwrap(step.from?.accountId, M10Error.InvalidTransaction())),
            publicName: unwrap(step.from?.publicName, M10Error.InvalidTransaction()),
            profileImageUrl: unwrap(step.from?.profileImageUrl, M10Error.InvalidTransaction()),
            code: unwrap(step.from?.code, M10Error.InvalidTransaction()),
            decimals: unwrap(step.from?.decimalPlaces, M10Error.InvalidTransaction()),
        };

        const to: ExpandedAccount = {
            id: AccountId.fromUint8Array(unwrap(step.to?.accountId, M10Error.InvalidTransaction())),
            publicName: unwrap(step.to?.publicName, M10Error.InvalidTransaction()),
            profileImageUrl: unwrap(step.to?.profileImageUrl, M10Error.InvalidTransaction()),
            code: unwrap(step.to?.code, M10Error.InvalidTransaction()),
            decimals: unwrap(step.to?.decimalPlaces, M10Error.InvalidTransaction()),
        };

        return new ExpandedTransferStep({
            from: from,
            to: to,
            amount: unwrap(raw.amount, M10Error.InvalidTransaction()),
            metadata: unwrap(raw.metadata, M10Error.InvalidTransaction()),
        });
    }
}


export interface IExpandedTransfer {
    txId: TxId;
    contextId: Uint8Array;
    timestamp: LongNumber;
    steps: ExpandedTransferStep[];
    success: boolean;
    status: TransferStatus;
}

export class ExpandedTransfer implements IExpandedTransfer {
    public txId: TxId;
    public contextId: Uint8Array;
    public timestamp: LongNumber;
    public steps: ExpandedTransferStep[];
    public success: boolean;
    public status: TransferStatus;


    public constructor(properties: IExpandedTransfer) {
        this.txId = properties.txId;
        this.contextId = properties.contextId;
        this.timestamp = properties.timestamp;
        this.steps = properties.steps;
        this.success = properties.success;
        this.status = properties.status;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     * @throws  {M10Error.InvalidAccountId}     if invalid account id was used
     */
    public static tryFrom(transfer: EnhancedTransfer): ExpandedTransfer {
        return new ExpandedTransfer({
            txId: transfer.transfer.txId,
            contextId: transfer.transfer.contextId,
            timestamp: transfer.transfer.timestamp,
            steps: transfer.enhanced_steps.map((step, index) => ExpandedTransferStep.tryFrom(step, transfer.transfer.transferSteps[index])),
            success: isNone(transfer.transfer.error),
            status: Transfer.getTransferStatus(transfer.transfer.state),
        });
    }
}
